metautils/lib/lb.c

```diff
@@ -29,1512 +29,121 @@ License along with this library.
-			(GDestroyNotify)grid_lb_clean);
-	glp->iterators = g_tree_new_full(metautils_strcmp3, NULL,
-			g_free, (GDestroyNotify)grid_lb_iterator_clean);
-	return glp;
-}
-
-void
-grid_lbpool_destroy(struct grid_lbpool_s *glp)
-{
-	if (!glp)
-		return;
-
-	g_rw_lock_writer_lock(&(glp->rwlock));
-
-	if (glp->iterators)
-		g_tree_destroy(glp->iterators);
-	if (glp->pools)
-		g_tree_destroy(glp->pools);
-
-	g_rw_lock_writer_unlock(&(glp->rwlock));
-	g_rw_lock_clear(&(glp->rwlock));
-
-	SLICE_FREE(struct grid_lbpool_s, glp);
-}
-
-static void
-_ensure(struct grid_lbpool_s *glp, const gchar *srvtype,
-		struct grid_lb_s **plb, struct grid_lb_iterator_s **pit)
-{
-	struct grid_lb_s *lb;
-	struct grid_lb_iterator_s *iterator;
-
-	if (!(lb = g_tree_lookup(glp->pools, srvtype))) {
-		lb = grid_lb_init(glp->ns, srvtype);
-		g_tree_insert(glp->pools, g_strdup(srvtype), lb);
-	}
-
-	if (!(iterator = g_tree_lookup(glp->iterators, srvtype))) {
-		iterator = grid_lb_iterator_weighted_round_robin(lb);
-		g_tree_insert(glp->iterators, g_strdup(srvtype), iterator);
-	}
-
-	if (pit)
-		*pit = iterator;
-	if (plb)
-		*plb = lb;
-}
-
-void
-grid_lbpool_configure_string(struct grid_lbpool_s *glp, const gchar *srvtype,
-		const gchar *cfg)
-{
-	EXTRA_ASSERT(glp != NULL);
-	EXTRA_ASSERT(srvtype != NULL);
-	EXTRA_ASSERT(cfg != NULL);
-	grid_lb_iterator_configure (grid_lbpool_ensure_iterator(glp, srvtype), cfg);
-}
-
-static void
-_configure_string(struct grid_lbpool_s *glp, const gchar *kv)
-{
-	GRID_TRACE("parsing : [%s]", kv);
-	const gchar *val = strchr(kv, '=');
-
-	if (!val || kv == val) {
-		GRID_DEBUG("Invalid format");
-		return;
-	}
-
-	gsize len = val - kv + 1;
-	gchar srvtype[len];
-	memcpy(srvtype, kv, val - kv);
-	srvtype[ len-1 ] = 0;
-	++ val;
-
-	grid_lbpool_configure_string(glp, srvtype, val);
-}
-
 void
-grid_lbpool_reconfigure(struct grid_lbpool_s *glp,
-		struct namespace_info_s *ni)
-{
-	if (!glp || !ni || !ni->options) {
-		GRID_DEBUG("Invalid parameter (%s)", __FUNCTION__);
-		return ;
-	}
-
-	GHashTableIter iter;
-	gpointer k, v;
-	g_hash_table_iter_init(&iter, ni->options);
-	while (g_hash_table_iter_next(&iter, &k, &v)) {
-		if (!g_str_has_prefix((gchar*)k, "lb."))
-			continue;
-		GString *str = g_string_new((gchar*)k + (sizeof("lb.") -1));
-		str = g_string_append_len(g_string_append(str, "="),
-				((char *)((GByteArray*)v)->data), ((GByteArray*)v)->len);
-		_configure_string(glp, str->str);
-		g_string_free(str, TRUE);
-	}
-}
-
-struct grid_lb_s *
-grid_lbpool_get_lb (struct grid_lbpool_s *glp, const gchar *srvtype)
+oio_lb_world__feed_service_info_list(struct oio_lb_world_s *lbw,
+		GSList *services)
 {
-	EXTRA_ASSERT (glp != NULL);
-	EXTRA_ASSERT (srvtype != NULL);
+	for (GSList *l = services; l; l = l->next) {
+		char slot_name[128] = {0};
+		struct service_info_s *srv = l->data;
+		/* Allocate item on the stack, it will be copied later */
+		struct oio_lb_item_s *item = g_alloca(sizeof(struct oio_lb_item_s)+128);
```

Comment:
Am I missing something or using g_alloca in a loop is the best way to run out of the stack ?